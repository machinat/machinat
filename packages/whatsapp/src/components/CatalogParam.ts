import { makePartSegment, PartSegment } from '@sociably/core/renderer';
import makeWhatsAppComponent from '../utils/makeWhatsAppComponent.js';
import { WhatsAppComponent } from '../types.js';

/** @category Props */
export type CatalogParamProps = {
  /**
   * Item SKU number. Labeled as Content ID in the Commerce Manager. The
   * thumbnail of this item will be used as the message's header image.
   */
  thumbnailProductRetailerId: string;
  /**
   * The 0-indexed position of the button. If the value is undefined, it's
   * decided by the order of params.
   */
  index?: number;
};

/**
 * Define the copied coupon code of a copy code button
 *
 * @category Component
 * @props {@link CatalogParamProps}
 */
export const CatalogParam: WhatsAppComponent<
  CatalogParamProps,
  PartSegment<{}>
> = makeWhatsAppComponent(function CatalogParam(node, path) {
  const { index, thumbnailProductRetailerId } = node.props;
  return [
    makePartSegment(node, path, {
      type: 'button',
      sub_type: 'catalog',
      index,
      parameters: [
        {
          type: 'action',
          action: {
            thumbnail_product_retailer_id: thumbnailProductRetailerId,
          },
        },
      ],
    }),
  ];
});
