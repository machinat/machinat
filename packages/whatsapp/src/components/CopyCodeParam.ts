import { makePartSegment, PartSegment } from '@sociably/core/renderer';
import makeWhatsAppComponent from '../utils/makeWhatsAppComponent.js';
import { WhatsAppComponent } from '../types.js';

/** @category Props */
export type CopyCodeParamProps = {
  /** The coupon code to be copied when the customer taps the button. */
  code: string;
  /**
   * The 0-indexed position of the button. If the value is undefined, it's
   * decided by the order of params.
   */
  index?: number;
};

/**
 * Define the coupon code of a copy code button
 *
 * @category Component
 * @props {@link CopyCodeParamProps}
 */
export const CopyCodeParam: WhatsAppComponent<
  CopyCodeParamProps,
  PartSegment<{}>
> = makeWhatsAppComponent(function CopyCodeParam(node, path) {
  const { index, code } = node.props;
  return [
    makePartSegment(node, path, {
      type: 'button',
      sub_type: 'copy_code',
      index,
      parameters: [
        {
          type: 'coupon_code',
          coupon_code: code,
        },
      ],
    }),
  ];
});
