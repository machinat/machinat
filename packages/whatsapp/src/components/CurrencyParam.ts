import { makePartSegment, PartSegment } from '@sociably/core/renderer';
import makeWhatsAppComponent from '../utils/makeWhatsAppComponent.js';
import { WhatsAppComponent } from '../types.js';

/** @category Props */
export type CurrencyParamProps = {
  /** Currency code as defined in ISO 4217 */
  code: string;
  /** Default text if localization fails */
  fallbackValue: string;
  /** Amount multiplied by 1000 */
  amount1000: number;
};

/**
 * Pass a currency parameter into the body or header of a customized template
 *
 * @category Component
 * @props {@link CurrencyParamProps}
 */
export const CurrencyParam: WhatsAppComponent<
  CurrencyParamProps,
  PartSegment<{ type: 'currency'; currency: {} }>
> = makeWhatsAppComponent(function CurrencyParam(node, path) {
  const { code, fallbackValue, amount1000 } = node.props;

  return [
    makePartSegment(node, path, {
      type: 'currency',
      currency: {
        code,
        fallback_value: fallbackValue,
        amount_1000: amount1000,
      },
    }),
  ];
});
