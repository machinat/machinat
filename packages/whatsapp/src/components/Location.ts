import { makeUnitSegment, UnitSegment } from '@sociably/core/renderer';
import makeWhatsAppComponent from '../utils/makeWhatsAppComponent.js';
import { WhatsAppSegmentValue, WhatsAppComponent } from '../types.js';

export type LocationProps = {
  /** Longitude of the location. */
  longitude: number;
  /** Latitude of the location. */
  latitude: number;
  /** Name of the location. */
  name?: string;
  /** Address of the location. Only displayed if name is present. */
  address?: string;
  /** Reply to the specified message */
  replyTo?: string;
};

/**
 * Send a location
 * @category Component
 * @props {@link LocationProps}
 */
export const Location: WhatsAppComponent<
  LocationProps,
  UnitSegment<WhatsAppSegmentValue>
> = makeWhatsAppComponent(function Location(node, path) {
  const { latitude, longitude, name, address, replyTo } = node.props;

  return [
    makeUnitSegment(node, path, {
      message: {
        type: 'location',
        location: { latitude, longitude, name, address },
        context: replyTo ? { message_id: replyTo } : undefined,
      },
    }),
  ];
});
