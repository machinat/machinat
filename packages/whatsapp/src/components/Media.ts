import { NativeElement, SociablyNode } from '@sociably/core';
import {
  makeUnitSegment,
  UnitSegment,
  InnerRenderFn,
} from '@sociably/core/renderer';
import makeWhatsAppComponent from '../utils/makeWhatsAppComponent';
import { WhatsAppSegmentValue, WhatsAppComponent } from '../types';

/**
 * @category Props
 */
export type MediaProps = {
  /** The uploaded media ID */
  mediaId?: string;
  /** The URL of the media to be sent */
  url?: string;
  /** The file content to be upload */
  fileData?: Buffer | NodeJS.ReadableStream;
  /** The content type of the file. Required when uploading using `file` */
  fileType?: string;
  /** Reply to the specified message */
  replyTo?: string;
  /** The asset tag for saving the created media and reusing it */
  assetTag?: string;
};

export type WithCaption = {
  /** Text to describes the media */
  caption?: SociablyNode;
};

export type WithFileName = {
  /**
   * Describes the filename for the specific document. The extension of the
   * filename will specify what format the document is displayed as in WhatsApp
   */
  fileName?: string;
};

const makeMediaComponent = (componentName: string, mediaType: string) => {
  const container = {
    [componentName]: async (
      node: NativeElement<MediaProps & WithCaption & WithFileName, any>,
      path: string,
      render: InnerRenderFn
    ) => {
      const {
        mediaId,
        url,
        fileData,
        fileType,
        caption,
        fileName,
        replyTo,
        assetTag,
      } = node.props;

      if (
        !(url || mediaId || fileData) ||
        (url && mediaId) ||
        (url && fileData) ||
        (fileData && mediaId)
      ) {
        throw new TypeError(
          'there should be exactly one of "url", "mediaId" or "fileData" prop'
        );
      }
      if (fileData && !fileType) {
        throw new TypeError(
          '"fileType" prop is required when using "fileData" source'
        );
      }

      const captionSegments = await render(caption, '.caption');
      if (captionSegments) {
        for (const seg of captionSegments) {
          if (seg.type !== 'text') {
            throw new TypeError(
              `"caption" prop should contain only textual content`
            );
          }
        }
      }

      return [
        makeUnitSegment(node, path, {
          message: {
            type: mediaType,
            [mediaType]: {
              id: mediaId,
              link: url,
              caption: captionSegments?.[0].value,
              filename: fileName,
            },
            context: replyTo ? { message_id: replyTo } : undefined,
          },
          mediaFile: fileData
            ? {
                data: fileData,
                type: fileType as string,
                info: { contentType: fileType, filename: fileName },
                assetTag,
              }
            : undefined,
        }),
      ];
    },
  };

  return makeWhatsAppComponent(container[componentName]);
};

/**
 * Send an audio
 * @category Component
 * @props {@link AudioProps}
 */
export const Audio: WhatsAppComponent<
  MediaProps,
  UnitSegment<WhatsAppSegmentValue>
> = makeMediaComponent('Audio', 'audio');

/**
 * Send a document
 * @category Component
 * @props {@link DocumentProps}
 */
export const Document: WhatsAppComponent<
  MediaProps & WithCaption & WithFileName,
  UnitSegment<WhatsAppSegmentValue>
> = makeMediaComponent('Document', 'document');

/**
 * Send an image
 * @category Component
 * @props {@link ImageProps}
 */
export const Image: WhatsAppComponent<
  MediaProps & WithCaption,
  UnitSegment<WhatsAppSegmentValue>
> = makeMediaComponent('Image', 'image');

/**
 * Send a sticker
 * @category Component
 * @props {@link StickerProps}
 */
export const Sticker: WhatsAppComponent<
  MediaProps,
  UnitSegment<WhatsAppSegmentValue>
> = makeMediaComponent('Sticker', 'sticker');

/**
 * Send a video
 * @category Component
 * @props {@link VideoProps}
 */
export const Video: WhatsAppComponent<
  MediaProps & WithCaption & WithFileName,
  UnitSegment<WhatsAppSegmentValue>
> = makeMediaComponent('Video', 'video');
