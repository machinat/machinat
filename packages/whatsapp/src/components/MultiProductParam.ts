import { makePartSegment, PartSegment } from '@sociably/core/renderer';
import makeWhatsAppComponent from '../utils/makeWhatsAppComponent.js';
import { WhatsAppComponent } from '../types.js';

type MultiPruductSction = {
  /** Section title text. Maximum 24 characters. Markdown is not supported. */
  title: string;
  /**
   * SKU number of the item you want to appear in the section. SKU numbers are
   * labeled as Content ID in the Commerce Manager. Supports up to 30 products
   * total, across all sections.
   */
  productItems: { productRetailerId: string }[];
};

/** @category Props */
export type MultiProductParamProps = {
  /**
   * Item SKU number. Labeled as Content ID in the Commerce Manager. The
   * thumbnail of this item will be used as the message's header image.
   */
  thumbnailProductRetailerId: string;
  /**
   * The 0-indexed position of the button. If the value is undefined, it's
   * decided by the order of params.
   */
  index?: number;
  /** You can define up to 10 sections. */
  sections: MultiPruductSction[];
};

/**
 * Define the copied coupon code of a copy code button
 *
 * @category Component
 * @props {@link MultiProductParamProps}
 */
export const MultiProductParam: WhatsAppComponent<
  MultiProductParamProps,
  PartSegment<{}>
> = makeWhatsAppComponent(function MultiProductParam(node, path) {
  const { index, thumbnailProductRetailerId } = node.props;
  return [
    makePartSegment(node, path, {
      type: 'button',
      sub_type: 'mpm',
      index,
      parameters: [
        {
          type: 'action',
          action: {
            thumbnail_product_retailer_id: thumbnailProductRetailerId,
          },
          sections: node.props.sections.map((section) => ({
            title: section.title,
            product_items: section.productItems.map((productItem) => ({
              product_retailer_id: productItem.productRetailerId,
            })),
          })),
        },
      ],
    }),
  ];
});
