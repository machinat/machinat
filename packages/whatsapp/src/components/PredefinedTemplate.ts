import { SociablyNode } from '@sociably/core';
import {
  makeUnitSegment,
  UnitSegment,
  IntermediateSegment,
} from '@sociably/core/renderer';
import { formatNode } from '@sociably/core/utils';
import makeWhatsAppComponent from '../utils/makeWhatsAppComponent.js';
import { WhatsAppSegmentValue, WhatsAppComponent } from '../types.js';

export type PredefinedTemplateProps = {
  /** Name of the template */
  name: string;
  /**
   * The code of the language or locale to use. Accepts both language and
   * language_locale formats (e.g., en and en_US). For all codes, see [Supported
   * Languages](https://developers.facebook.com/docs/whatsapp/api/messages/message-templates#supported-languages)
   */
  languageCode: string;
  /**
   * Header parameters. For text header, use only `TextParameter` or
   * `CurrencyParameter`. For media header, use exactly one `Image`, `Video` or
   * `Document` element.
   */
  headerParams?: SociablyNode;
  /**
   * Body parameters. It should contain only `TextParameter`,
   * `CurrencyParameter` and `DateTimeParameter`.
   */
  bodyParams?: SociablyNode;
  /**
   * Button parameters. Use only `QuickReplyParameter` or `UrlButtonParameter`
   * depends on the pre-defined button type.
   */
  buttonParams?: SociablyNode;
  /** Reply to the specified message */
  replyTo?: string;
};

const HEADER_MEDIA_TYPES = ['image', 'document', 'video'];
const TEXT_PARAM_TYPES = ['text', 'currency', 'date_time'];
const BUTTON_PARAM_TYPES = ['url', 'quick_reply'];

const checkTextParams = (segments: IntermediateSegment<unknown>[]) => {
  for (const seg of segments) {
    if (!TEXT_PARAM_TYPES.includes(seg.value.type)) {
      throw new TypeError(
        `${formatNode(seg.node)} is not a valid text parameter`,
      );
    }
  }
};

const getValue = ({ value }: IntermediateSegment<unknown>) => value;

/**
 * Send a pre-defined custom template. WhatsApp message templates are specific
 * message formats that businesses use to send out notifications or customer
 * care messages to people that have opted in to notifications. Messages can
 * include appointment reminders, shipping information, issue resolution or
 * payment updates.
 *
 * Before sending a template message, you need to create a template. See [Create
 * Message Templates for Your WhatsApp Business
 * Account](https://developers.facebook.com/micro_site/url/?click_from_context_menu=true&country=TW&destination=https%3A%2F%2Fwww.facebook.com%2Fbusiness%2Fhelp%2F2055875911147364&event_type=click&last_nav_impression_id=11qPRPLhwqt8j8VOZ&max_percent_page_viewed=9&max_viewport_height_px=859&max_viewport_width_px=1512&orig_http_referrer=https%3A%2F%2Fdevelopers.facebook.com%2Fdocs%2Fwhatsapp%2Fcloud-api%2Fguides%2Fsend-message-templates&orig_request_uri=https%3A%2F%2Fdevelopers.facebook.com%2Fajax%2Fdocs%2Fnav%2F%3Fpath1%3Dwhatsapp%26path2%3Dcloud-api%26path3%3Dguides%26path4%3Dsend-message-templates&region=apac&scrolled=true&session_id=1AzR1OsntMXnFwSoY&site=developers)
 * for more information. If your account is not verified yet, you can use
 * [pre-approved
 * templates](https://developers.facebook.com/micro_site/url/?click_from_context_menu=true&country=TW&destination=https%3A%2F%2Fwww.facebook.com%2Fbusiness%2Fhelp%2F722393685250070&event_type=click&last_nav_impression_id=11qPRPLhwqt8j8VOZ&max_percent_page_viewed=9&max_viewport_height_px=859&max_viewport_width_px=1512&orig_http_referrer=https%3A%2F%2Fdevelopers.facebook.com%2Fdocs%2Fwhatsapp%2Fcloud-api%2Fguides%2Fsend-message-templates&orig_request_uri=https%3A%2F%2Fdevelopers.facebook.com%2Fajax%2Fdocs%2Fnav%2F%3Fpath1%3Dwhatsapp%26path2%3Dcloud-api%26path3%3Dguides%26path4%3Dsend-message-templates&region=apac&scrolled=true&session_id=1AzR1OsntMXnFwSoY&site=developers).
 *
 * @category Component
 * @props {@link PredefinedTemplateProps}
 */
export const PredefinedTemplate: WhatsAppComponent<
  PredefinedTemplateProps,
  UnitSegment<WhatsAppSegmentValue>
> = makeWhatsAppComponent(
  async function PredefinedTemplate(node, path, render) {
    const {
      name,
      languageCode,
      headerParams,
      bodyParams,
      buttonParams,
      replyTo,
    } = node.props;

    const [headerSegments, bodySegments, buttonsSegments] = await Promise.all([
      render(headerParams, '.headerParams'),
      render(bodyParams, '.bodyParams'),
      render(buttonParams, '.buttonParams'),
    ]);

    const components: unknown[] = [];

    if (headerSegments) {
      if ('message' in headerSegments[0].value) {
        // media based header
        const messageValue = headerSegments[0].value.message;
        const messageType = messageValue.type;

        if (!HEADER_MEDIA_TYPES.includes(messageType)) {
          throw new TypeError(
            `${formatNode(headerSegments[0].node)} is not a valid parameter`,
          );
        }
        if (headerSegments.length > 1) {
          throw new TypeError(`"headerParams" prop contain more than 1 media`);
        }

        components.push({
          type: 'header',
          parameters: [
            {
              type: messageType,
              [messageType]: messageValue[messageType],
            },
          ],
        });
      } else {
        // text based header
        checkTextParams(headerSegments);

        components.push({
          type: 'header',
          parameters: headerSegments.map(getValue),
        });
      }
    }
    if (bodySegments) {
      checkTextParams(bodySegments);

      components.push({
        type: 'body',
        parameters: bodySegments.map(getValue),
      });
    }
    if (buttonsSegments) {
      buttonsSegments.forEach(({ node: buttonNode, value }, idx) => {
        if (!BUTTON_PARAM_TYPES.includes(value.type)) {
          throw new TypeError(
            `${formatNode(buttonNode)} is not a valid button parameter`,
          );
        }

        components.push({
          type: 'button',
          sub_type: value.type,
          index: typeof value.index === 'undefined' ? idx : value.index,
          parameters: [value.parameter],
        });
      });
    }

    return [
      makeUnitSegment(node, path, {
        message: {
          type: 'template',
          template: {
            name,
            language: {
              policy: 'deterministic',
              code: languageCode,
            },
            components,
          },
          context: replyTo ? { message_id: replyTo } : undefined,
        },
      }),
    ];
  },
);
