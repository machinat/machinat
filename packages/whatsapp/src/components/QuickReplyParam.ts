import { makePartSegment, PartSegment } from '@sociably/core/renderer';
import makeWhatsAppComponent from '../utils/makeWhatsAppComponent.js';
import { WhatsAppComponent } from '../types.js';

/**
 * @category Props
 */
export type QuickReplyParamProps = {
  /**
   * Developer-defined payload that is returned when the button is clicked in
   * addition to the display text on the button.
   */
  payload: string;
  /**
   * The 0-indexed position of the button. If the value is undefined, it's
   * decided by the order of params.
   */
  index?: number;
};

/**
 * Define the payload of a quick reply button
 * @category Component
 * @props {@link QuickReplyParamProps}
 */
export const QuickReplyParam: WhatsAppComponent<
  QuickReplyParamProps,
  PartSegment<{}>
> = makeWhatsAppComponent(function QuickReplyParam(node, path) {
  const { index, payload } = node.props;
  return [
    makePartSegment(node, path, {
      type: 'quick_reply',
      index,
      parameter: {
        type: 'payload',
        payload,
      },
    }),
  ];
});
