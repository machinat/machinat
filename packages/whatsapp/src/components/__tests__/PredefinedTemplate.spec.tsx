import Sociably from '@sociably/core';
import { isNativeType } from '@sociably/core/utils';
import { PredefinedTemplate } from '../PredefinedTemplate.js';
import { Image, Audio } from '../Media.js';
import { Location } from '../Location.js';
import { TextParam } from '../TextParam.js';
import { CurrencyParam } from '../CurrencyParam.js';
import { QuickReplyParam } from '../QuickReplyParam.js';
import { UrlButtonParam } from '../UrlButtonParam.js';
import { CopyCodeParam } from '../CopyCodeParam.js';
import { CatalogParam } from '../CatalogParam.js';
import { MultiProductParam } from '../MultiProductParam.js';
import { renderUnitElement } from './utils.js';

it('is a valid Component', () => {
  expect(isNativeType(<PredefinedTemplate name="" language="" />)).toBe(true);
  expect(PredefinedTemplate.$$platform).toBe('whatsapp');
  expect(PredefinedTemplate.$$name).toBe('PredefinedTemplate');
});

test('rendering value', async () => {
  await expect(
    renderUnitElement(<PredefinedTemplate name="FOO" language="en" />),
  ).resolves.toMatchSnapshot();
  await expect(
    renderUnitElement(
      <PredefinedTemplate
        name="MY_TEMPLATE"
        language="ja"
        headerParams={<Location latitude={25.033} longitude={121.565} />}
        bodyParams="Hello World"
        buttonParams={
          <>
            <UrlButtonParam index={1} urlPostfix="/baz" />
            <CopyCodeParam index={2} code="HELLO_WORLD" />
          </>
        }
      />,
    ),
  ).resolves.toMatchSnapshot();
  await expect(
    renderUnitElement(
      <PredefinedTemplate
        name="MY_TEMPLATE"
        language="CA"
        headerParams={<Image url="http://foo.bar/baz.jpg" />}
        bodyParams={
          <>
            <TextParam>
              <b>hello</b>
            </TextParam>
            <TextParam>
              <i>world</i>
            </TextParam>
          </>
        }
        buttonParams={
          <>
            <QuickReplyParam payload="FOO" />
            <QuickReplyParam payload="BAR" />
            <QuickReplyParam payload="BAZ" />
          </>
        }
        replyTo="REPLY_TO_MESSAGE_ID"
      />,
    ),
  ).resolves.toMatchSnapshot();
  await expect(
    renderUnitElement(
      <PredefinedTemplate
        name="MY_TEMPLATE"
        language="ca"
        headerParams="Hello Header"
        bodyParams={
          <>
            <b>Hello</b> <i>Body</i>
          </>
        }
        buttonParams={
          <CatalogParam thumbnailProductRetailerId="_PRODUCT_ID_" />
        }
        replyTo="REPLY_TO_MESSAGE_ID"
      />,
    ),
  ).resolves.toMatchSnapshot();
  await expect(
    renderUnitElement(
      <PredefinedTemplate
        name="MY_TEMPLATE"
        language="en"
        headerParams={
          <>
            <TextParam>
              <b>FOO</b>
            </TextParam>
            <CurrencyParam code="USD" amount1000={123} fallbackValue="US$123" />
          </>
        }
        bodyParams={
          <>
            <TextParam>
              <i>BAR</i>
            </TextParam>
            <CurrencyParam code="TWD" amount1000={999} fallbackValue="TW$999" />
          </>
        }
        buttonParams={
          <MultiProductParam
            thumbnailProductRetailerId="_PRODUCT_ID_"
            sections={[
              {
                title: 'Section 1',
                productItems: [
                  { productRetailerId: '_PRODUCT_ID_1_' },
                  { productRetailerId: '_PRODUCT_ID_2_' },
                ],
              },
              {
                title: 'Section 2',
                productItems: [
                  { productRetailerId: '_PRODUCT_ID_3_' },
                  { productRetailerId: '_PRODUCT_ID_4_' },
                  { productRetailerId: '_PRODUCT_ID_5_' },
                ],
              },
            ]}
          />
        }
      />,
    ),
  ).resolves.toMatchSnapshot();
});

it('throw if invalid params received', async () => {
  await expect(
    renderUnitElement(
      <PredefinedTemplate
        name="FOO"
        language="en"
        headerParams={<Audio mediaId="123" />}
      />,
    ),
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"<Audio /> is not a valid header parameter"`,
  );
  await expect(
    renderUnitElement(
      <PredefinedTemplate
        name="FOO"
        language="en"
        headerParams={
          <>
            <Image mediaId="123" />
            <Image mediaId="123" />
          </>
        }
      />,
    ),
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `""headerParams" prop contain more than 1 attachment parameter node"`,
  );
  await expect(
    renderUnitElement(
      <PredefinedTemplate
        name="FOO"
        language="en"
        headerParams={<QuickReplyParam payload="" />}
      />,
    ),
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"<QuickReplyParam /> is not a valid text parameter"`,
  );
  await expect(
    renderUnitElement(
      <PredefinedTemplate
        name="FOO"
        language="en"
        bodyParams={<QuickReplyParam payload="" />}
      />,
    ),
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"<QuickReplyParam /> is not a valid text parameter"`,
  );
  await expect(
    renderUnitElement(
      <PredefinedTemplate
        name="FOO"
        language="en"
        bodyParams={<Image mediaId="123" />}
      />,
    ),
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"<Image /> is not a valid text parameter"`,
  );
  await expect(
    renderUnitElement(
      <PredefinedTemplate
        name="FOO"
        language="en"
        buttonParams={<TextParam>_</TextParam>}
      />,
    ),
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"<TextParam /> is not a valid button parameter"`,
  );
  await expect(
    renderUnitElement(
      <PredefinedTemplate
        name="FOO"
        language="en"
        buttonParams={<Image mediaId="123" />}
      />,
    ),
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"<Image /> is not a valid button parameter"`,
  );
});
