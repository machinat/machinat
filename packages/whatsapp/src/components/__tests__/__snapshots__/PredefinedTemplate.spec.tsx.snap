// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering value 1`] = `
[
  {
    "node": <PredefinedTemplate
      language="en"
      name="FOO"
    />,
    "path": "$",
    "type": "unit",
    "value": {
      "message": {
        "context": undefined,
        "template": {
          "components": [],
          "language": {
            "code": "en",
          },
          "name": "FOO",
        },
        "type": "template",
      },
    },
  },
]
`;

exports[`rendering value 2`] = `
[
  {
    "node": <PredefinedTemplate
      bodyParams="Hello World"
      buttonParams={
        <Sociably.Fragment>
          <UrlButtonParam
            index={1}
            urlPostfix="/baz"
          />
          <CopyCodeParam
            code="HELLO_WORLD"
            index={2}
          />
        </Sociably.Fragment>
      }
      headerParams={
        <Location
          latitude={25.033}
          longitude={121.565}
        />
      }
      language="ja"
      name="MY_TEMPLATE"
    />,
    "path": "$",
    "type": "unit",
    "value": {
      "message": {
        "context": undefined,
        "template": {
          "components": [
            {
              "parameters": [
                {
                  "location": {
                    "address": undefined,
                    "latitude": 25.033,
                    "longitude": 121.565,
                    "name": undefined,
                  },
                  "type": "location",
                },
              ],
              "type": "header",
            },
            {
              "parameters": [
                {
                  "text": "Hello World",
                  "type": "text",
                },
              ],
              "type": "body",
            },
            {
              "index": 1,
              "parameters": [
                {
                  "text": "/baz",
                  "type": "text",
                },
              ],
              "sub_type": "url",
              "type": "button",
            },
            {
              "index": 2,
              "parameters": [
                {
                  "coupon_code": "HELLO_WORLD",
                  "type": "coupon_code",
                },
              ],
              "sub_type": "copy_code",
              "type": "button",
            },
          ],
          "language": {
            "code": "ja",
          },
          "name": "MY_TEMPLATE",
        },
        "type": "template",
      },
    },
  },
]
`;

exports[`rendering value 3`] = `
[
  {
    "node": <PredefinedTemplate
      bodyParams={
        <Sociably.Fragment>
          <TextParam>
            <b>
              hello
            </b>
          </TextParam>
          <TextParam>
            <i>
              world
            </i>
          </TextParam>
        </Sociably.Fragment>
      }
      buttonParams={
        <Sociably.Fragment>
          <QuickReplyParam
            payload="FOO"
          />
          <QuickReplyParam
            payload="BAR"
          />
          <QuickReplyParam
            payload="BAZ"
          />
        </Sociably.Fragment>
      }
      headerParams={
        <Image
          url="http://foo.bar/baz.jpg"
        />
      }
      language="CA"
      name="MY_TEMPLATE"
      replyTo="REPLY_TO_MESSAGE_ID"
    />,
    "path": "$",
    "type": "unit",
    "value": {
      "message": {
        "context": {
          "message_id": "REPLY_TO_MESSAGE_ID",
        },
        "template": {
          "components": [
            {
              "parameters": [
                {
                  "image": {
                    "caption": undefined,
                    "filename": undefined,
                    "id": undefined,
                    "link": "http://foo.bar/baz.jpg",
                  },
                  "type": "image",
                },
              ],
              "type": "header",
            },
            {
              "parameters": [
                {
                  "text": "*hello*",
                  "type": "text",
                },
                {
                  "text": "_world_",
                  "type": "text",
                },
              ],
              "type": "body",
            },
            {
              "index": 0,
              "parameters": [
                {
                  "payload": "FOO",
                  "type": "payload",
                },
              ],
              "sub_type": "quick_reply",
              "type": "button",
            },
            {
              "index": 1,
              "parameters": [
                {
                  "payload": "BAR",
                  "type": "payload",
                },
              ],
              "sub_type": "quick_reply",
              "type": "button",
            },
            {
              "index": 2,
              "parameters": [
                {
                  "payload": "BAZ",
                  "type": "payload",
                },
              ],
              "sub_type": "quick_reply",
              "type": "button",
            },
          ],
          "language": {
            "code": "CA",
          },
          "name": "MY_TEMPLATE",
        },
        "type": "template",
      },
    },
  },
]
`;

exports[`rendering value 4`] = `
[
  {
    "node": <PredefinedTemplate
      bodyParams={
        <Sociably.Fragment>
          <b>
            Hello
          </b>
           
          <i>
            Body
          </i>
        </Sociably.Fragment>
      }
      buttonParams={
        <CatalogParam
          thumbnailProductRetailerId="_PRODUCT_ID_"
        />
      }
      headerParams="Hello Header"
      language="ca"
      name="MY_TEMPLATE"
      replyTo="REPLY_TO_MESSAGE_ID"
    />,
    "path": "$",
    "type": "unit",
    "value": {
      "message": {
        "context": {
          "message_id": "REPLY_TO_MESSAGE_ID",
        },
        "template": {
          "components": [
            {
              "parameters": [
                {
                  "text": "Hello Header",
                  "type": "text",
                },
              ],
              "type": "header",
            },
            {
              "parameters": [
                {
                  "text": "*Hello* _Body_",
                  "type": "text",
                },
              ],
              "type": "body",
            },
            {
              "index": 0,
              "parameters": [
                {
                  "action": {
                    "thumbnail_product_retailer_id": "_PRODUCT_ID_",
                  },
                  "type": "action",
                },
              ],
              "sub_type": "catalog",
              "type": "button",
            },
          ],
          "language": {
            "code": "ca",
          },
          "name": "MY_TEMPLATE",
        },
        "type": "template",
      },
    },
  },
]
`;

exports[`rendering value 5`] = `
[
  {
    "node": <PredefinedTemplate
      bodyParams={
        <Sociably.Fragment>
          <TextParam>
            <i>
              BAR
            </i>
          </TextParam>
          <CurrencyParam
            amount1000={999}
            code="TWD"
            fallbackValue="TW$999"
          />
        </Sociably.Fragment>
      }
      buttonParams={
        <MultiProductParam
          sections={
            [
              {
                "productItems": [
                  {
                    "productRetailerId": "_PRODUCT_ID_1_",
                  },
                  {
                    "productRetailerId": "_PRODUCT_ID_2_",
                  },
                ],
                "title": "Section 1",
              },
              {
                "productItems": [
                  {
                    "productRetailerId": "_PRODUCT_ID_3_",
                  },
                  {
                    "productRetailerId": "_PRODUCT_ID_4_",
                  },
                  {
                    "productRetailerId": "_PRODUCT_ID_5_",
                  },
                ],
                "title": "Section 2",
              },
            ]
          }
          thumbnailProductRetailerId="_PRODUCT_ID_"
        />
      }
      headerParams={
        <Sociably.Fragment>
          <TextParam>
            <b>
              FOO
            </b>
          </TextParam>
          <CurrencyParam
            amount1000={123}
            code="USD"
            fallbackValue="US$123"
          />
        </Sociably.Fragment>
      }
      language="en"
      name="MY_TEMPLATE"
    />,
    "path": "$",
    "type": "unit",
    "value": {
      "message": {
        "context": undefined,
        "template": {
          "components": [
            {
              "parameters": [
                {
                  "text": "*FOO*",
                  "type": "text",
                },
                {
                  "currency": {
                    "amount_1000": 123,
                    "code": "USD",
                    "fallback_value": "US$123",
                  },
                  "type": "currency",
                },
              ],
              "type": "header",
            },
            {
              "parameters": [
                {
                  "text": "_BAR_",
                  "type": "text",
                },
                {
                  "currency": {
                    "amount_1000": 999,
                    "code": "TWD",
                    "fallback_value": "TW$999",
                  },
                  "type": "currency",
                },
              ],
              "type": "body",
            },
            {
              "index": 0,
              "parameters": [
                {
                  "action": {
                    "thumbnail_product_retailer_id": "_PRODUCT_ID_",
                  },
                  "sections": [
                    {
                      "product_items": [
                        {
                          "product_retailer_id": "_PRODUCT_ID_1_",
                        },
                        {
                          "product_retailer_id": "_PRODUCT_ID_2_",
                        },
                      ],
                      "title": "Section 1",
                    },
                    {
                      "product_items": [
                        {
                          "product_retailer_id": "_PRODUCT_ID_3_",
                        },
                        {
                          "product_retailer_id": "_PRODUCT_ID_4_",
                        },
                        {
                          "product_retailer_id": "_PRODUCT_ID_5_",
                        },
                      ],
                      "title": "Section 2",
                    },
                  ],
                  "type": "action",
                },
              ],
              "sub_type": "mpm",
              "type": "button",
            },
          ],
          "language": {
            "code": "en",
          },
          "name": "MY_TEMPLATE",
        },
        "type": "template",
      },
    },
  },
]
`;
